src/megaapi_impl.cpp

```diff
@@ -26925,51 +27223,117 @@ void MegaFolderDownloadController::downloadFolderNode(MegaNode *node, LocalPath&
 
     if (!children)
     {
-        LOG_err << "Child nodes not found: " << localpath.toPath(*client->fsaccess);
+        LOG_err << "Child nodes not found: " << megaApi->LocalPathToPath(localpath);
         recursive--;
         mLastError = API_ENOENT;
         mIncompleteTransfers++;
-        checkCompletion();
         return;
     }
 
     for (int i = 0; i < children->size(); i++)
     {
         MegaNode *child = children->get(i);
-
-        ScopedLengthRestore restoreLen(localpath);
-        localpath.appendWithSeparator(LocalPath::fromName(child->getName(), *client->fsaccess, fsType), true);
-
-        string utf8path = localpath.toPath(*client->fsaccess);
-
         if (child->getType() == MegaNode::TYPE_FILE)
         {
-            pendingTransfers++;
-            megaApi->startDownload(false, child, utf8path.c_str(), tag, transfer->getAppData(), this);
+            // Add child node to vector in mLocalTree at index we have stored it's localPath
+            ::mega::unique_ptr<MegaNode> childNode (child->copy());
+            mLocalTree.at(index).childrenNodes.push_back(std::move(childNode));
+            mPendingFilesToProcess++;
         }
         else
         {
-            downloadFolderNode(child, localpath, fsType);
+            ScopedLengthRestore restoreLen(localpath);
+            localpath.appendWithSeparator(megaApi->getLocalPathFromName(child->getName(), fsType), true);
+            scanFolder(child, localpath, fsType);
         }
     }
 
     recursive--;
-    checkCompletion();
     if (deleteChildren)
     {
         delete children;
     }
 }
 
-void MegaFolderDownloadController::checkCompletion()
+void MegaFolderDownloadController::createFolder()
 {
-    if (!cancelled && !recursive && !pendingTransfers)
+    // Create all local directories in one shot
+    auto it = mLocalTree.begin();
+    while (it != mLocalTree.end())
     {
-        LOG_debug << "Folder download finished - " << transfer->getTransferredBytes() << " of " << transfer->getTotalBytes();
-        transfer->setState(MegaTransfer::STATE_COMPLETED);
-        transfer->setLastError(&mLastError);
-        DBTableTransactionCommitter committer(client->tctable);
-        megaApi->fireOnTransferFinish(transfer, make_unique<MegaErrorPrivate>(!mIncompleteTransfers ? API_OK : API_EINCOMPLETE), committer);
+        LocalPath &localpath = it->localPath;
+        mLastError = megaApi->createLocalFolder(localpath);
+        if (mLastError.getErrorCode() && mLastError.getErrorCode() != API_EEXIST)
+        {
+            mIncompleteTransfers++;
+            it = mLocalTree.erase(it); // remove all it's children nodes
+            mPendingFilesToProcess -= it->childrenNodes.size();
```

Comment:
It looks like this might be a silent failure - shouldn't we inform the user of a failure so that they can decide how to address the problem?